### PR TITLE
#18217: Added new variant any selected

### PR DESCRIPTION
**Pull request by Chumva**

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java [hunk 1]
```diff
@@ -259,18 +256,18 @@ public void onDeviceDisconnect(@NonNull AbstractDevice<?> device) {
...
 		if (getTargetFragment() instanceof SelectDeviceListener) {
-			((SelectDeviceListener) getTargetFragment()).selectNewDevice(device, widgetDataFieldType);
+			((SelectDeviceListener) getTargetFragment()).selectNewDevice(deviceId, widgetDataFieldType);
 			requireActivity().onBackPressed();
 		}
 	}
 
 	enum States {
-		NO_BLUETOOTH, NOTHING_FOUND, CONTENT
+		NO_BLUETOOTH, CONTENT
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java:267]: Can we guarantee that we always find something?

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java [hunk 2]
```diff
@@ -204,7 +199,7 @@ private void updatePairedSensorsList() {
 		List<AbstractDevice<?>> filteredDevices = plugin.getPairedDevicesByWidgetType(widgetDataFieldType);
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java:199]: unused devices list?

File: OsmAnd/res/layout/fragment_select_external_sensors.xml [hunk 3]
```diff
@@ -32,10 +32,6 @@
 		android:layout_height="wrap_content"
 		osmand:layout_behavior="com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior">
 
-		<include
-			android:id="@+id/empty_view"
-			layout="@layout/select_external_sensors_empty" />
```
Comment (Chumva) [inline OsmAnd/res/layout/fragment_select_external_sensors.xml:37]: Probably this layout should be deleted if it is not used anywhere

File: OsmAnd/src/net/osmand/plus/plugins/monitoring/MonitoringSettingsFragment.java [hunk 4]
```diff
@@ -302,10 +305,21 @@ private List<String> getLinkedSensorNames() {
...
+					boolean sensorLinked = false;
+					if (!ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY.equals(deviceId)) {
+						if (sensorsPlugin.getDevice(deviceId) != null) {
+							sensorLinked = true;
+						}
+					} else {
+						if (sensorsPlugin.getDevice(widgetDataField) != null) {
+							sensorLinked = true;
+						}
+					}
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/monitoring/MonitoringSettingsFragment.java:319]: looks strange

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java [hunk 5]
```diff
@@ -72,7 +73,8 @@ public class ExternalSensorsPlugin extends OsmandPlugin {
 	public final CommonPreference<String> POWER_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> HEART_RATE_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> TEMPERATURE_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
-	public final static String DENY_WRITE_SENSOR_DATA_TO_TRACK_KEY = "deny_write_sensor_data";
+	public final static String ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY = "any_connected_device_write_sensor_data_to_track_key";
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java:76]: Will there be any issues with this change if we save it as a key in preference? Is migration needed?

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/adapters/DevicesForWidgetAdapter.java [hunk 6]
```diff
@@ -30,43 +41,68 @@ public void setDeviceId(@Nullable String selectedDeviceId) {
...
 		}
 		holder.selectionMark.setVisibility(View.VISIBLE);
 	}
 
+	@Override
+	public int getItemViewType(int position) {
+		if (position == 0) {
+			return withNoneVariant ? NONE_ITEM_TYPE : ANY_CONNECTED_ITEM_TYPE;
+		} else if (position == 1) {
+			return withNoneVariant ? ANY_CONNECTED_ITEM_TYPE : DEVICE_ITEM_TYPE;
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/adapters/DevicesForWidgetAdapter.java:76]: Looks complicated, probably better to add constants to items in the adapter, similar to [ImportTracksAdapter](https://github.com/osmandapp/Osmand/blob/a174c86ee49cca0bb6af2895cd2bb0afda0468dc/OsmAnd/src/net/osmand/plus/importfiles/ui/ImportTracksAdapter.java#L128)

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/adapters/DevicesForWidgetAdapter.java [hunk 7]
```diff
@@ -30,43 +43,66 @@ public void setDeviceId(@Nullable String selectedDeviceId) {
...
```
Comment (Chumva) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/adapters/DevicesForWidgetAdapter.java:80]: Looks complicated, probably better to add constants to items in the adapter, similar to [ImportTracksAdapter](https://github.com/osmandapp/Osmand/blob/a174c86ee49cca0bb6af2895cd2bb0afda0468dc/OsmAnd/src/net/osmand/plus/importfiles/ui/ImportTracksAdapter.java#L128)

File: OsmAnd/src/net/osmand/plus/views/mapwidgets/configure/settings/SensorWidgetSettingFragment.java [hunk 8]
```diff
@@ -64,39 +64,51 @@ protected void initParams(@NonNull Bundle bundle) {
...
+			if (sensorWidget != null) {
+				if (sensorWidget.getWidgetSensor() != null) {
+					connectedDevice = sensorWidget.getWidgetSensor().getDevice();
+				} else {
+					connectedDevice = plugin.getDevice(sensorWidget.getFieldType());
+				}
+			}
+			StringBuilder builder = new StringBuilder(getString(R.string.any_connected));
+			if (connectedDevice != null) {
+				builder.append(": ").append(connectedDevice.getName());
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/views/mapwidgets/configure/settings/SensorWidgetSettingFragment.java:109]: Localize

File: OsmAnd/src/net/osmand/plus/views/mapwidgets/configure/settings/SensorWidgetSettingFragment.java [hunk 9]
```diff
@@ -64,39 +64,51 @@ protected void initParams(@NonNull Bundle bundle) {
...
 		if (sourceDevice == null) {
 			deviceIcon.setImageDrawable(uiUtils.getIcon(R.drawable.ic_action_sensor, nightMode));
-			((TextView) view.findViewById(R.id.device_name)).setText(R.string.shared_string_none);
+			AbstractDevice<?> connectedDevice = null;
+			if (sensorWidget != null) {
+				if (sensorWidget.getWidgetSensor() != null) {
+					connectedDevice = sensorWidget.getWidgetSensor().getDevice();
+				} else {
+					connectedDevice = plugin.getDevice(sensorWidget.getFieldType());
+				}
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/views/mapwidgets/configure/settings/SensorWidgetSettingFragment.java:105]: Device should be returned by widget

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java [hunk 10]
```diff
@@ -44,35 +57,25 @@
...
-	private ScanDevicesListener scanDevicesListener;
 	private final OsmandSettings settings;
+	private final DevicesHelper devicesHelper;
 
 	public final CommonPreference<String> SPEED_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> CADENCE_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> POWER_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> HEART_RATE_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
 	public final CommonPreference<String> TEMPERATURE_SENSOR_WRITE_TO_TRACK_DEVICE_ID;
-	public final static String DENY_WRITE_SENSOR_DATA_TO_TRACK_KEY = "deny_write_sensor_data";
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java:75]: Does migration needed?

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java [hunk 11]
```diff
@@ -44,35 +57,25 @@
...
-import static net.osmand.plus.plugins.externalsensors.devices.sensors.SensorWidgetDataFieldType.BIKE_DISTANCE;
-import static net.osmand.plus.plugins.externalsensors.devices.sensors.SensorWidgetDataFieldType.BIKE_POWER;
-import static net.osmand.plus.plugins.externalsensors.devices.sensors.SensorWidgetDataFieldType.BIKE_SPEED;
-import static net.osmand.plus.plugins.externalsensors.devices.sensors.SensorWidgetDataFieldType.HEART_RATE;
-import static net.osmand.plus.plugins.externalsensors.devices.sensors.SensorWidgetDataFieldType.TEMPERATURE;
-
 public class ExternalSensorsPlugin extends OsmandPlugin {
 	private static final Log LOG = PlatformUtil.getLog(ExternalSensorsPlugin.class);
 	private static final int DEVICES_SEARCH_TIMEOUT = 10000;
+	public static final String ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY = "any_connected_device_write_sensor_data_to_track_key";
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java:66]: This inner constant should be private. Rename to "any_device". New method like `shouldWriteAnyDeviceToTrack(dataType)` should be added to plugin.

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java (hunk 10 above)
Comment (vshcherb) [reply OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java:75]: I think no

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java (hunk 1 above)
Comment (Corwin-Kh) [reply OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java:269]: if nothing found, we have None variant. it was one of the point in task to remove empty view

File: OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java (hunk 5 above)
Comment (Corwin-Kh) [reply OsmAnd/src/net/osmand/plus/plugins/externalsensors/ExternalSensorsPlugin.java]: migration should not be needed. there should be checks if we can find device with such deviceId

**Pull request merged**